### ISSUE TITLE
[Windows] - failed to add windows node in airgap setup

### ISSUE THREAD

**Issue by jiaqiluo** [rancher/rancher#22966]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug	

**Steps to reproduce (least amount of steps as possible):**
In the airgap environment, try to add a Windows worker node to the custom cluster. I copy and paste the command from Rancher UI to the powerShell in the windows node, and here is the output:
```
PS C:\> docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
PS C:\>
PS C:\>
PS C:\> PowerShell -NoLogo -NonInteractive -Command "& {docker run -v c:/:c:/host ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5 bootstrap --server https://jiaqi-ag-lb-e4f7e0724eaa18d0.elb.us-west-1.amazonaws.com --token gcjwzjcmt8bzn27tmxxk658dlx9hfwff4625clw85wrxl6jkkw765q --ca-checksum 78b36df18d1a6c24d5a85c8f616b36178ed6f5bbe376886b8a648d3ab8acbe6f --worker | iex}"
Unable to find image 'ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5' locally
v2.3.0-rc5: Pulling from rancher/rancher-agent
3f81696d6765: Already exists
446dc6511ddc: Already exists
ffa747c13622: Already exists
b62f66ce2ff6: Already exists
6829b790015b: Already exists
1a5508aa7d2f: Already exists
96bfb26b3151: Already exists
bff9d71854d0: Already exists
352f0bc2c9ca: Already exists
c71c11ca485e: Already exists
df69c1d7f186: Already exists
67cf58d99f12: Already exists
9d65a1522e84: Already exists
da1fff6b47b4: Already exists
b7667626c68e: Already exists
89983e886988: Already exists
237778543dd4: Already exists
fffd633f9048: Already exists
ed4a098899e3: Already exists
5f910aa6cc76: Already exists
9fdd73c027b2: Already exists
8dba85e44c23: Already exists
00c27c1e790a: Already exists
641171d05d6c: Already exists
Digest: sha256:2023a4c9858258f5fa565f98b5d569d50c12c74a4b1e80ec6b0eb381a8679e11
Status: Downloaded newer image for ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5
INFO: Detecting running permission ...
INFO: Detecting host CPU reservation ...
INFO: Detecting host RAM reservation ...
INFO: Detecting host DISK reservation ...
INFO: Detecting host system locale ...
INFO: Detecting host Docker name pipe existing ...
INFO: Detecting host Docker release version ...
INFO: Detecting host network interface ...
INFO: Configuring host Docker startup mode to automatical ...
INFO: Enabling host msiscsi service to support iscsi storage ...
WARN[2019-09-19T18:27:14Z] Service is not stopped, going to stop it
WARN[2019-09-19T18:27:17Z] Service is not stopped, going to stop it
FATA[2019-09-19T18:27:20Z] failed to register service: could not query status: The service has not been started.
Unable to find image 'rancher/rancher-agent:v2.3.0-rc5' locally
C:\Program Files\Docker\docker.exe: Error response from daemon: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
See 'C:\Program Files\Docker\docker.exe run --help'.
PS C:\> docker images
REPOSITORY                                                                    TAG                 IMAGE ID            CREATED             SIZE
ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent   v2.3.0-rc5          08436fc3bfaa        37 hours ago        561MB
PS C:\
```

I can fix the `cannot stop service error` by restarting docker daemon. But although the image `ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5` is pulled from my registry, docker still tries to start `rancher/rancher-agent:v2.3.0-rc5`.

**Comment by bmdepesa** [comment]
https://github.com/rancher/rancher/blob/543d3822bdc80bf675a57c2a7ca047b79e738f4d/package/windows/bootstrap.ps1#L272

I think it's possible `AGENT_IMAGE` does not have the private registry prefix

**Comment by jiaqiluo** [comment]
Thanks @bmdepesa  to point out that we can set the Env variable in the docker command line. So I run
```
PowerShell -NoLogo -NonInteractive -Command "& {docker run -v c:/:c:/host -e AGENT_IMAGE=ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5  ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5 bootstrap --server https://jiaqi-ag-lb-e4f7e0724eaa18d0.elb.us-west-1.amazonaws.com --token gcjwzjcmt8bzn27tmxxk658dlx9hfwff4625clw85wrxl6jkkw765q --ca-checksum 78b36df18d1a6c24d5a85c8f616b36178ed6f5bbe376886b8a648d3ab8acbe6f --worker | iex}"
```

here is the output:
 
```
PS C:\> PowerShell -NoLogo -NonInteractive -Command "& {docker run -v c:/:c:/host -e AGENT_IMAGE=ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5  ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5 bootstrap --server https://jiaqi-ag-lb-e4f7e0724eaa18d0.elb.us-west-1.amazonaws.com --token gcjwzjcmt8bzn27tmxxk658dlx9hfwff4625clw85wrxl6jkkw765q --ca-checksum 78b36df18d1a6c24d5a85c8f616b36178ed6f5bbe376886b8a648d3ab8acbe6f --worker | iex}"
INFO: Detecting running permission ...
INFO: Detecting host CPU reservation ...
INFO: Detecting host RAM reservation ...
INFO: Detecting host DISK reservation ...
INFO: Detecting host system locale ...
INFO: Detecting host Docker name pipe existing ...
INFO: Detecting host Docker release version ...
INFO: Detecting host network interface ...
INFO: Configuring host Docker startup mode to automatical ...
INFO: Enabling host msiscsi service to support iscsi storage ...
WARN[2019-09-19T19:40:10Z] Service is not stopped, going to stop it
WARN[2019-09-19T19:40:14Z] Service is not stopped, going to stop it
FATA[2019-09-19T19:40:17Z] failed to register service: could not query status: The service has not been started.
b0abd8db4bc033350c7037be7ecb516d901ac890a6be835328a18b24f5763c2c
PS C:\> docker ps
CONTAINER ID        IMAGE                                                                                    COMMAND                  CREATED             STATUS              PORTS               NAMES
b0abd8db4bc0        ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/rancher-agent:v2.3.0-rc5   "pwsh -NoLogo -NonIn…"   7 seconds ago       Up 5 seconds                            sad_austin
``` 

But the node is stuck in `registering` state in Rancher, and the log from the `service-sidekick` container in the node shows this
```
E0919 19:58:10.842316    5520 main.go:294] Error registering network: failed to acquire lease: node "ec2amaz-emeekib" not found
I0919 19:58:10.842316    5520 main.go:373] Stopping shutdownHandler...
INFO: Start flanneld with: --kubeconfig-file=c:\etc\kubernetes\ssl\kubecfg-kube-node.yaml --ip-masq --kube-subnet-mgr --iptables-forward-rules=false --iface=172.31.26.132
I0919 19:58:21.856651     172 main.go:457] Searching for interface using 172.31.26.132
I0919 19:58:22.007554     172 main.go:534] Using interface with name Ethernet and address 172.31.26.132
I0919 19:58:22.007554     172 main.go:551] Defaulting external address to interface address (172.31.26.132)
I0919 19:58:22.019599     172 kube.go:126] Waiting 10m0s for node controller to sync
I0919 19:58:22.019599     172 kube.go:309] Starting kube subnet manager
I0919 19:58:23.031476     172 kube.go:133] Node controller sync successful
I0919 19:58:23.031595     172 main.go:249] Created subnet manager: Kubernetes Subnet Manager - ec2amaz-emeekib
I0919 19:58:23.031595     172 main.go:252] Installing signal handlers
I0919 19:58:23.031595     172 main.go:393] Found network config - Backend type: vxlan
I0919 19:58:23.031595     172 vxlan_windows.go:116] VXLAN config: Name=vxlan0 VNI=4096 Port=4789 GBP=false DirectRouting=false
E0919 19:58:23.042664     172 main.go:294] Error registering network: failed to acquire lease: node "ec2amaz-emeekib" not found
I0919 19:58:23.042664     172 main.go:373] Stopping shutdownHandler...
```

**Comment by jiaqiluo** [comment]
So we have 2 issues here:
 1. we need to fix the environment variable `AGENT_IMAGE` such that it can catch the private registry prefix
 2. even we pass `AGENT_IMAGE` in the docker run command and workaround issue 1, and the node is added to the cluster (as shown in the UI ), the state is stuck in registering

**Comment by jiaqiluo** [comment]
In the windows node 

cannot pull this image
 `ec2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/hyperkube:v1.15.3-rancher3`
but can pull this one 
 `c2-13-56-165-173.us-west-1.compute.amazonaws.com:443/rancher/hyperkube:v1.15.3-rancher3-windows-1903`

**Comment by thxCode** [comment]
For `#1`, I would try to find a way to catch the private registry image into AGENT_IMAGE, but it might be hard as Windows doesn't has a CGroup record like Linux: https://github.com/rancher/rancher/blob/7bf6786c2dab19d539b7e6afa34557f5d502cefd/package/share-root.sh#L5. so I prefer to your workaround at the moment.

For `#2`, you need to clean up the pre-adding host before re-add it, there is a cleanup script on `c:\etc\rancher\cleanup.ps1`.

**Comment by thxCode** [comment]
After discussing with @loganhz , in order to solve https://github.com/rancher/rancher/issues/22966#issuecomment-533383440 `#1`, we will open a new PR to fix it: the UI would allow users to input the `private registry address` when adding host to the custom cluster. cc @n313893254 .

**Comment by jiaqiluo** [comment]
The bug fix is validated with Rancher:v2.3.0-rc7 release 

Steps:
- Follow the instruction to prepare the private registry: https://github.com/rancher/docs/issues/1742#issuecomment-532553582 
- Follow the Rancher documentation to install Rancher in airgap HA mode
- add a custom cluster with both Linux and Windows node 

Result:
- the Windows node is added to the cluster successfully
